### PR TITLE
chore(storage-image-processing-api): initial extension build

### PR DESCRIPTION
**Pull request by Salakar**
Introducing a Storage image processing extension which provides an extensive image processing API with 30+ image mutation operations from resizing and colour manipulation to overlaying text and compositing images.

The core image processing logic in the implementation is portable and therefore can be used in a variety of ways, e.g., 
 - reacting to a storage event to automatically process an image (operations could be defined via a metadata key)
 - adding image operations to Firestore and listening to a Firestore collection that would trigger these operations to be applied and output
 - etc

but for this implementation, initially, interfacing with the image processing is provided as a URL safe HTTP GET API via a Cloud Function (and could be used for embedding e.g., ``).

Accessing GCS bucket objects is done via tokenated download urls, e.g., `getDownloadUrl()` on Firebase SDKs or providing a path to a _public_ object on the bucket, e.g., `/foo/bar.png`. URLs & paths must be URL encoded.

## Process Image URL Schema

The URL schema to apply operations is built into url parts separated by `/`, each segment is then in the following format;

```
/<operationName>~<optionName>:<optionValue>/
```

 - A single operation and it's options are defined per url segment. Multiple options are separated by `~`. 
 - Option values are automatically coerced into their appropriate types, e.g. `'5'` as a string would become `5` as a number, or `'[1,2,3,4]'` as a string would result in a JS array of numbers. 
 - Truthy `boolean` options can just specify the `<optionName>` only to have the value default to `true` (think flags). 
 - Operations that have no options or only have optional options can just specify the operation without options, e.g.`/negate/`.
 - An `input` operation must be the first operation specified and is required.
 - An `output` operation must be the last operation specified and is required.

Having the URL schema as url segments vs query parameters allows us to easily have operation ordering and support using multiple operations of the same type in a chain of operations. Additionally makes options clearer that they are for a specific operation. 

### Example

**URL:**

```
/process/input~type:gcs~source:%2Fassets%2Fhexagons.jpeg/resize~width:500~height:500~fit:cover~position:left%20bottom/extend~all:20~background:tomato/extend~all:20~background:lightgreen/composite~top:20~left:26~tile~blend:screen~input:%2Fassets%2Finvertase.png/text~strokeWidth:4~top:75~left:75~font:95px%20tahoma~strokeColor:SeaGreen~textAlign:center~value:Hello%0DWorld~gravity:northwest~backgroundColor:pink~textColor:MediumSpringGreen~padding:15~borderWidth:5~borderColor:black/output~format:png
```

**Output:**

image

## Supported operations

 - [input](https://github.com/invertase/firebase-extensions/blob/storage-image-processing-api/extensions/storage-image-processing-api/functions/src/operations/input.ts)
 - [output](https://github.com/invertase/firebase-extensions/blob/storage-image-processing-api/extensions/storage-image-processing-api/functions/src/operations/output.ts)
 - [affine](https://github.com/invertase/firebase-extensions/blob/storage-image-processing-api/extensions/storage-image-processing-api/functions/src/operations/affine.ts)
 - [blur](https://github.com/invertase/firebase-extensions/blob/storage-image-processing-api/extensions/storage-image-processing-api/functions/src/operations/blur.ts)
 - [clahe](https://github.com/invertase/firebase-extensions/blob/storage-image-processing-api/extensions/storage-image-processing-api/functions/src/operations/clahe.ts)
 - [composite](https://github.com/invertase/firebase-extensions/blob/storage-image-processing-api/extensions/storage-image-processing-api/functions/src/operations/composite.ts) 
   - experimental, needs some TLC around retrieving input sources
 - [extend](https://github.com/invertase/firebase-extensions/blob/storage-image-processing-api/extensions/storage-image-processing-api/functions/src/operations/extend.ts)
 - [extract](https://github.com/invertase/firebase-extensions/blob/storage-image-processing-api/extensions/storage-image-processing-api/functions/src/operations/extract.ts) 
   - a.k.a crop
 - [flatten](https://github.com/invertase/firebase-extensions/blob/storage-image-processing-api/extensions/storage-image-processing-api/functions/src/operations/flatten.ts)
   - e.g., can be used to remove transparency
 - [flip](https://github.com/invertase/firebase-extensions/blob/storage-image-processing-api/extensions/storage-image-processing-api/functions/src/operations/flip.ts)
 - [flop](https://github.com/invertase/firebase-extensions/blob/storage-image-processing-api/extensions/storage-image-processing-api/functions/src/operations/flop.ts)
 - [gamma](https://github.com/invertase/firebase-extensions/blob/storage-image-processing-api/extensions/storage-image-processing-api/functions/src/operations/gamma.ts)
 - [median](https://github.com/invertase/firebase-extensions/blob/storage-image-processing-api/extensions/storage-image-processing-api/functions/src/operations/median.ts)
 - [negate](https://github.com/invertase/firebase-extensions/blob/storage-image-processing-api/extensions/storage-image-processing-api/functions/src/operations/negate.ts)
 - [normalize](https://github.com/invertase/firebase-extensions/blob/storage-image-processing-api/extensions/storage-image-processing-api/functions/src/operations/normalize.ts)
 - [resize](https://github.com/invertase/firebase-extensions/blob/storage-image-processing-api/extensions/storage-image-processing-api/functions/src/operations/resize.ts)
 - [rotate](https://github.com/invertase/firebase-extensions/blob/storage-image-processing-api/extensions/storage-image-processing-api/functions/src/operations/rotate.ts)
 - [sharpen](https://github.com/invertase/firebase-extensions/blob/storage-image-processing-api/extensions/storage-image-processing-api/functions/src/operations/sharpen.ts)
 - [text](https://github.com/invertase/firebase-extensions/blob/storage-image-processing-api/extensions/storage-image-processing-api/functions/src/operations/text.ts)
   - custom implementation since Sharp doesn't have a direct api for this - uses the composition API to overlay a generated SVG. This same kind of logic with overlaying SVGs could be used to do things like gradients & shapes
 - [trim](https://github.com/invertase/firebase-extensions/blob/storage-image-processing-api/extensions/storage-image-processing-api/functions/src/operations/trim.ts)
 - [convolve](https://github.com/invertase/firebase-extensions/blob/storage-image-processing-api/extensions/storage-image-processing-api/functions/src/operations/convolve.ts)
 - [threshold](https://github.com/invertase/firebase-extensions/blob/storage-image-processing-api/extensions/storage-image-processing-api/functions/src/operations/threshold.ts)
 - [boolean](https://github.com/invertase/firebase-extensions/blob/storage-image-processing-api/extensions/storage-image-processing-api/functions/src/operations/boolean.ts)
 - [linear](https://github.com/invertase/firebase-extensions/blob/storage-image-processing-api/extensions/storage-image-processing-api/functions/src/operations/linear.ts)
 - [recomb](https://github.com/invertase/firebase-extensions/blob/storage-image-processing-api/extensions/storage-image-processing-api/functions/src/operations/recomb.ts)
 - [modulate](https://github.com/invertase/firebase-extensions/blob/storage-image-processing-api/extensions/storage-image-processing-api/functions/src/operations/modulate.ts)
 - [tint](https://github.com/invertase/firebase-extensions/blob/storage-image-processing-api/extensions/storage-image-processing-api/functions/src/operations/tint.ts)
 - [grayscale](https://github.com/invertase/firebase-extensions/blob/storage-image-processing-api/extensions/storage-image-processing-api/functions/src/operations/grayscale.ts)
 - [colorspace](https://github.com/invertase/firebase-extensions/blob/storage-image-processing-api/extensions/storage-image-processing-api/functions/src/operations/colorspace.ts)

### Operations pending implementation

Not blocked or difficult by any means, just needs time invested to implement. These will be added in a future PR.

 - [ ] removeAlpha
 - [ ] ensureAlpha
 - [ ] extractChannel
 - [ ] joinChannel
 - [ ] bandbool

---

## TODO

 - [x] Finish up `composite` - currently sourcing from local files for testing purposes, should source from same sources that the `input` operation supports.
 - [x] Add remaining `output` operation format specific options, e.g., https://github.com/invertase/firebase-extensions/blob/storage-image-processing-api/extensions/storage-image-processing-api/functions/src/operations/output.ts#L32
 - [x] Add all available operation types (Currently Prs)
 - [x] Tests
 - [ ] Documentation
 - [ ] Input from GCS bucket path
 - [ ] Investigate possible alternatives to output to a bucket rather than returning to the requesting client (maybe a POST api)

### CODE REVIEW

**Comment by Salakar** [comment]
Could do cartoon filters with `median` maybe; 

image

**Comment by Salakar** [comment]
`Negate`:

image

**Comment by Salakar** [comment]
Example of me mutating the `Melos` logo;

**Original:**
logo


**Mutated:**

```
http://localhost:3001/process/input~type:url~url:https%3A%2F%2Fraw.githubusercontent.com%2Finvertase%2Fmelos%2Fmaster%2Fdocs%2Fassets%2Flogo-dark.png/flip/flop/median~size:55/extend~right:300~top:100~left:100/text~value:elos~top:350~left:475~font:110px%20arial/output~format:jpeg
```

image

File: extensions/storage-image-processing-api/extension.yaml [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+      API requests with input urls or paths that are not inside this bucket will be
+      dropped.
+    type: string
+    example: my-project-12345.appspot.com
+    validationRegex: ^([0-9a-z_.-]*)$
+    validationErrorMessage: Invalid storage bucket
+    default: ${STORAGE_BUCKET}
+    required: true
+
+  - param: CORS_ORIGIN_ALLOW_LIST
```
Comment (jhuleatt) [inline extensions/storage-image-processing-api/extension.yaml:110]: Great idea! This will definitely be important to highlight

**Comment by jhuleatt** [comment]
Very cool! Overall, this looks like a huge improvement. A couple of thoughts:

**Cache-control headers**: It would be especially valuable to web developers if cache headers could be set. `storage-resize-images` makes this configurable via [a param](https://github.com/firebase/extensions/blob/d49a7ed3d2b3b74746e02bc759ce7560d5c1b258/storage-resize-images/extension.yaml#L212-L221), but maybe this could be done with a query string or added to the URL schema?

**Saving resized images in Cloud Storage**: I remember you mentioned that Sharp is fast enough that this may not be worth the Storage cost (or time cost of saving the image into Storage the first time?). But I'm wondering if this function could return a [permanent redirect](https://developer.mozilla.org/en-US/docs/Web/HTTP/Redirections#permanent_redirections) to the processed image's Cloud Storage downloadUrl, and in that way the browser should fetch the image directly from the Storage bucket next time, avoiding function execution entirely. I haven't totally thought this through, but maybe something to look into?

**Building the URL**: There are [well-established packages](https://www.npmjs.com/package/query-string) for parsing/building query strings, but since this URL schema will be new, it's probably worth providing at least a code sample of how to build a URL, and maybe even a package if you're up for maintaining that across platforms. I know this extension is still early, but something to think about later.

File: extensions/storage-image-processing-api/functions/src/index.ts [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+
+    res.writeHead(200, headers);
+    res.end(data);
+  }),
+);
+
+// Express requires the function to have 4 arguments for a handler
+// to be treated as an error handler.
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+app.use(function (error, req, res, next) {
```
Comment (jhuleatt) [inline extensions/storage-image-processing-api/functions/src/index.ts:125]: nitpick: May help to name this function so its purpose is immediately clear

```suggestion
app.use(function handleError(error, req, res, next) {
```

Comment (Salakar) [reply extensions/storage-image-processing-api/functions/src/index.ts:125]: Great suggestion, thanks!

**Comment by Salakar** [comment]
Thanks @jhuleatt for the feedback 🎉 

> **Cache-control headers**: It would be especially valuable to web developers if cache headers could be set so that the browser doesn't need to make a network call every time it re-visits a page. `storage-resize-images` makes this configurable via [a param](https://github.com/firebase/extensions/blob/d49a7ed3d2b3b74746e02bc759ce7560d5c1b258/storage-resize-images/extension.yaml#L212-L221), but maybe this could be done with a query string or added to the URL schema?
> An added benefit would be that if someone really wants their images served quickly and publicly, they could set this extension up with Hosting (via [rewrites](https://firebase.google.com/docs/hosting/functions#direct-requests-to-function)), and tune the CDN cache settings through the URL schema/query params.

This sounds extremely useful, will look into it.


> **Saving resized images in Cloud Storage**: I remember you mentioned that Sharp is fast enough that this may not be worth the Storage cost (or time cost of saving the image into Storage the first time?). But I'm wondering if this function could return a [permanent redirect](https://developer.mozilla.org/en-US/docs/Web/HTTP/Redirections#permanent_redirections) to the processed image's Cloud Storage downloadUrl, and in that way the browser should fetch the image directly from the Storage bucket next time, avoiding function execution entirely. I haven't totally thought this through, but maybe something to look into?

Definitely already looking into this, will update here once I have something for this. My initial testing was that sharp was indeed faster than a network request to GCS to fetch a pre-cached item. Sharp in most cases with several operations applied averaged around ~30ms (excluding fetching inputs) vs fetching a cached image from GCS taking around 200-400ms. However this was excluding the initial fetch of the input image, so in most cases if there's an input image then having cache works out to be slightly faster. With cached results we could also be smarter and for example do redirects and cache control headers - like you've mentioned above.


> **Building the URL**: There are [well-established packages](https://www.npmjs.com/package/query-string) for parsing/building query strings, but since this URL schema will be new, it's probably worth providing at least a code sample of how to build a URL, and maybe even a package if you're up for maintaining that across platforms. I know this extension is still early, but something to think about later.

I posted some initial reasonings on the URL structure vs say query in the original PR description above, I'd say the code to produce a valid url is fairly minimal and could be added as client samples or just documentation (something I'd definitely add) - we could even ship the validators as a client api for TS/JS if needed since they are fairly well separated. I need to do some more thinking on the URL setup overall so will come back to this but am open to suggestions

**Pull request merged**